src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -1007,5 +1006,74 @@ void M()
                     //         Func<string, string, string> func = (x!!, y) => x;
                     Diagnostic(ErrorCode.ERR_FeatureInPreview, "!!").WithArguments("parameter null-checking").WithLocation(7, 47));
         }
+
+        [Fact]
+        public void TestNullCheckedParameterUpdatesFlowState1()
+        {
+            var source =
+@"
+#nullable enable
+
+class Program
+{
+    string M(string? s!!) // 1
+    {
+        return s;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (6,22): warning CS8995: Nullable type 'string?' is null-checked and will throw if null.
+                //     string M(string? s!!) // 1
+                Diagnostic(ErrorCode.WRN_NullCheckingOnNullableType, "s").WithArguments("string?").WithLocation(6, 22));
+        }
+
+        [Fact]
+        public void TestNullCheckedParameterUpdatesFlowState2()
+        {
+            var source =
+@"
+#nullable enable
+
+class Program
+{
+    int M(int? x!!) // 1
+    {
+        return x.Value;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (6,16): warning CS8995: Nullable type 'int?' is null-checked and will throw if null.
+                //     int M(int? x!!) // 1
+                Diagnostic(ErrorCode.WRN_NullCheckingOnNullableType, "x").WithArguments("int?").WithLocation(6, 16));
+        }
+
+        [Theory]
+        [InlineData("!!")]
+        [InlineData("")]
+        public void TestNullCheckedParameterDoesNotAffectNullableVarianceChecks(string nullCheckAnnotation)
+        {
+            var source =
+@"
+#nullable enable
+
+class Base
+{
+    #pragma warning disable 8995 // WRN_NullCheckingOnNullableType
+    public virtual void M(string? s" + nullCheckAnnotation + @") { }
+}
+
+class Derived : Base
+{
+    public override void M(string s) { } // 1
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (12,26): warning CS8765: Nullability of type of parameter 's' doesn't match overridden member (possibly because of nullability attributes).
+                //     public override void M(string s) { } // 1
+                Diagnostic(ErrorCode.WRN_TopLevelNullabilityMismatchInParameterTypeOnOverride, "M").WithArguments("s").WithLocation(12, 26)
+            );
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Minor: Would this test be easier to read if both cases were included together rather than using `[Theory]`?
```csharp
class Base
{
    #pragma warning disable 8995 // WRN_NullCheckingOnNullableType
    public virtual void M1(string? s) { }
    public virtual void M2(string? s!!) { }
}

class Derived : Base
{
    public override void M1(string s) { }
    public override void M2(string s) { }
}
```

Reply:
Yes.